### ISSUE TITLE
Other way to write SLO than ErrorQuery and TotalQuery

### ISSUE THREAD

**Issue by kyledakid** [slok/sloth#320]
Hi, Xabier

I have [Prometheus plugin](https://plugins.jenkins.io/prometheus/) installed in my Jenkins, which is using [DropWizard metrics](https://metrics.dropwizard.io/4.2.0/). The plugin exports the metric `jenkins_job_total_duration` that already includes `quantile` label. Eg: jenkins_job_total_duration{quantile="0.999"} = 300ms, meaning that 99.9% of Jenkins jobs are having the duration 300ms. However, there's no way to know how many jobs are having that 300ms duration. It's not like the Bucket implementation of Prometheus, where we apply `histogram_quantile` function and `le` label to know the duration and how many jobs.

In this case, I can't write the SLO "99.9% Jobs should have duration less than 300ms" in the meta of "ErrorQuery and TotalQuery". Because `ms` is not `jobs` to be divided, they're not in the same unit.

So my question is: Is there another way to write SLO in this case ? Does Sloth support another SLO declaration without ErrorQuery and TotalQuery ? Something like [RED framework](https://www.weave.works/blog/the-red-method-key-metrics-for-microservices-architecture/). Would love to know your opinion to define SLO for this case.

I love your work.
Thanks, Xabier.

**Comment by jimethn** [comment]
I am having the same problem. I don't think it's possible to define an SLO using a [summary metric](https://prometheus.io/docs/concepts/metric_types/#summary) (the kind with the quantile label) in SLOth.

Sloth needs you to include a `[{{.window}}]` variable somewhere in your query so it can define the [multiwindow multi-burn-rate alerts](https://sre.google/workbook/alerting-on-slos/). MWMBR requires the metric be evaluated over 5m, 6h, 3d, etc. For a summary metric the evaluation window is pre-computed on the client side and there is no way to adjust it in your query.

**Comment by jimethn** [comment]
While _conceptually_ it's possible to define an SLO using a quantile, I think _practically_ it cuts out a lot of options.

I guess you could do something like `max(max_over_time(jenkins_job_total_duration{quantile="0.9"}[{{.window}}]))`. Max is one of the few ways you can meaningfully [aggregate a quantile](https://prometheus.io/docs/practices/histograms/#quantiles). This will give you the slowest p90 over time.

The next problem is there's no way to say what percentage of requests met the SLI -- a quantile is all-or-nothing over a given scrape interval. You need your query to return a value between 0 and 1, where 0 is no errors and 1 is nothing but errors. So if your SLI is "the p90 of jobs will complete in less than 1 minute" you'll have to define that like

```
sli:
  raw:
    errorRatioQuery: |
      clamp(
        max(max_over_time(jenkins_job_total_duration{quantile="0.9"}[{{.window}}])) > 60,
        0, 1
      ) or vector(0)
```

Basically, if it's `> 60` we count it as an error period (1) and otherwise it's a success period (0). I am still not sure how will this will work with MWMBR alerting though, because if your SLI is anywhere close to your actual performance, those longer evaluation windows like 3d will probably end up being 1 most of the time.

Sorry to talk your ear off. This is something I've been struggling with the past few days, and I've pretty much come to the conclusion that I need to put a proxy in place to gather histograms, because summaries just don't cut it. I'd love if someone could come in and tell me where I'm going wrong.

**Comment by kyledakid** [comment]
Thanks for your opinion @jimethn

**Comment by slok** [comment]
Hello @kyledakid 

Thanks for the kind words!

@jimethn is right. the SLO model that Sloth implements, which is the one from Google's framework, is based on events. So you would need to think of good and bad (boolean form) events. _You need to make a question_ to your metrics, in the sense of _it's is good or bad?_

So in the case of summaries, you should always say what is the max latency on your `pXX` (normally p99) you would accept as good or bad. Example:

> My Jenkins job latency should be less than 500ms

That would imply that  you need to convert that to good bad events in a window timeframe:

- Bad: Count all the Jenkins job that have a duration greater than 500ms as bad
- Total: Count all the Jenkins jobs.

A way of implementing this would be using `_over_time`, count and subqueries:

- Bad: `sum_over_time((count(jenkins_job_total_duration{quantile="0.9"} > 0.5))[{{.window}}:]) OR on() vector(0)`
- Total: `sum_over_time((count(jenkins_job_total_duration{quantile="0.9"}))[{{.window}}:]) OR on() vector(0)`

You can take a look at my [home wifi's SLO example](https://sloth.dev/examples/default/home-wifi/). It's based on metrics that give the "wifi satisfaction" as a percent ratio for each connection. I assume that a connection with less than 75% satisfaction is bad and each connection would be a single event.

Hope this helps a bit :smile:

**Comment by kyledakid** [comment]
No, Xabier. This doesn't help a bit. This helps a ton, absolutely 😁 

Thank you very much @jimethn and @slok 

The quesiton can be closed
